src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs

```diff
@@ -396,9 +396,18 @@ private Endpoint CreateRejectionEndpoint(IEnumerable<string> httpMethods)
 
         private static bool ContainsHttpMethod(List<string> httpMethods, string httpMethod)
         {
-            for (var i = 0; i < httpMethods.Count; i++)
+            var methods = httpMethods.ToArray();
+            for (var i = 0; i < methods.Length; i++)
             {
-                if (string.Equals(httpMethods[i], httpMethod, StringComparison.OrdinalIgnoreCase))
+                if (object.ReferenceEquals(methods[i], httpMethod))
+                {
+                    return true;
+                }
+            }
+
+            for (var i = 0; i < methods.Length; i++)
+            {
+                if (StringComparer.OrdinalIgnoreCase.Equals(methods[i], httpMethod))
```

Comment:
Might as well use HttpMethods.Equals here since you're already off the fast path?